---
title: | 
  | Script for analysing bulk ATAC-seq of reprogramming intermediate
author: |
  | Fernando J. Rossello
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    toc: true
    toc_depth: 2
    toc_float: 
      collapsed: false
fontsize: 12pt
pagetitle: "bkATAC"
---

# Preamble
## Things to note
- All input files can be downloaded from http://hrpi.ddnetbio.com/
- All input files are assumed to be in the data folder
- All genome browser tracks can be explored at http://hrpi.ddnetbio.com at the _bulk ATAC-seq iPSC reprog. section_.

## Load required libraries.

```{r session_setup, include = TRUE, echo=TRUE, results='hide', message=FALSE, warning=FALSE}
library(dplyr)
library(edgeR)
library(irlba)
library(sjstats)
library(Mfuzz)
library(patchwork)
library(data.table)
library(ggplot2)
library(ggrepel)
# rm(list=ls())
```
## Define colour palettes.

```{r color_labels}

colorMedia = c("black","darkorange","blue","darkolivegreen2","red2","pink2")
names(colorMedia) = c("D0, 3, 7","Primed","t2iLGoY","5iLAF","NHSM","RSeT")
shapeTime = c(0,1,2,15,16,17,18)
names(shapeTime) = c("D0","D3","D7","D13","D21","P3","P10")
```

# Exploratory Analysis - PCA.
## Import read counts of consensus peak set.
```{r import_read_counts}
setwd("~/projects/Polo_Group/Ethan_Liu/Ethan_Liu_ATAC-seq_Human_Reprogramming_human_samples/analysis/R/scripts/github")
peak.counts <-
  read.table(file = "data/bkATAC_consensus_peak_set_counts.txt",
             header = T,
             stringsAsFactors = F)
```

### Reorder samples and save peak information.
```{r reorder_samples}
order <- colnames(peak.counts)[7:ncol(peak.counts)]
order <- order[c(1, 2, 11:14, 3, 5, 7, 9, 19, 21, 15, 17, 4, 6, 8, 10, 20, 22, 16, 18)]
order <- factor(order, levels = order)

peak.info <- colnames(peak.counts)[1:6]
```

### Reorder peak counts table.

```{r prepare_counts_table}

peak.counts <- peak.counts[, c(peak.info, levels(order))]

# Change Geneid column name to PeakID

colnames(peak.counts)[1] <- "PeakID"
rownames(peak.counts) <- peak.counts$PeakID
```

## Within and between library normalization.

### Quantile normalise log2(RPKM + 1)

```{r quantile_log2}

peaks.fpkm <-
  rpkm(peak.counts[, 7:ncol(peak.counts)],
       log = F, gene.length = peak.counts$Length)

# Keep peaks with more than 5 FPKMs in at least 2 samples for exploratory analysis.

keep <- rowSums(peaks.fpkm > 5) >= 2

peaks.fpkm <- peaks.fpkm[keep, ]
peaks.fpkm <- log2(peaks.fpkm + 1)
peaks.fpkm <- normalizeQuantiles(peaks.fpkm)
```

## PCA 

```{r pca}
# Prepare annotation table.
pca.plot.annotation <- data.table(sample_id = colnames(peaks.fpkm))

pca.plot.annotation$sample_labels <- c(
  paste(rep("FM_", 3), rep("D", 3), rep(c(0, 3, 7), each = 2), rep(c("_R1", "_R2")), sep = ""),
  paste(rep("PM_", 2), rep("D", 2), rep(c(13, 21), each = 2), rep(c("_R1", "_R2")), sep = ""),
  paste(rep("PM_", 2), rep("P", 2), rep(c(3, 10), each = 2), rep(c("_R1", "_R2")), sep = ""),
  paste(rep("NM_", 2), rep("D", 2), rep(c(13, 21), each = 2), rep(c("_R1", "_R2")), sep = ""),
  paste(rep("NM_", 2), rep("P", 2), rep(c(3, 10), each = 2), rep(c("_R1", "_R2")), sep = "")
)

pca.plot.annotation$media = "D0, 3, 7"
pca.plot.annotation[grep("PM", sample_labels)]$media = "Primed"
pca.plot.annotation[grep("NM", sample_labels)]$media = "t2iLGoY"
pca.plot.annotation$media = factor(pca.plot.annotation$media, 
                           levels = c("D0, 3, 7", "Primed", "t2iLGoY"))
pca.plot.annotation$timept = tstrsplit(pca.plot.annotation$sample_labels, "_")[[2]]
pca.plot.annotation$timept = factor(pca.plot.annotation$timept, 
                            levels = unique(pca.plot.annotation$timept))

set.seed(1234)
pca <- prcomp_irlba(t(peaks.fpkm))

pca.plot.annotation$PC1 = pca$x[,1]
pca.plot.annotation$PC2 = pca$x[,2]
pca.plot.annotation$PC3 = pca$x[,3]

ggplot(pca.plot.annotation, aes(PC1, PC2, color = media, label = timept)) +
  geom_point(size = 3) + geom_text_repel(size = 5) +
  scale_color_manual(values = colorMedia[1:3]) +
  scale_shape_manual(values = c(16, 15)) + theme_classic(base_size = 24)
  
ggplot(pca.plot.annotation, aes(PC1, PC3, color = media, label = timept)) +
  geom_point(size = 3) + geom_text_repel(size = 5) +
  scale_color_manual(values = colorMedia[1:3]) +
  scale_shape_manual(values = c(16, 15)) + theme_classic(base_size = 24)


```

# Fuzzy clustering.

## Aggregate sample counts means and calculate RPKM.

```{r aggregate_sample_counts_rpkm}
# Prepare conditions labels.

conditions.labels <-
  c(
  "D0",
  "D3",
  "D7",
  "D13.*Primed",
  "D21.*Primed",
  "P3.*Primed",
  "P10.*Primed",
  "D13.*Smith",
  "D21.*Smith",
  "P3.*Smith",
  "P10.*Smith"
  )

# Aggregate peak counts.
aggregated.peak.counts <-
  sapply(conditions.labels, function(x)
    rowSums(peak.counts[, grep(x,
                               colnames(peak.counts),
                               value = TRUE)]))

colnames(aggregated.peak.counts) <-
  colnames(aggregated.peak.counts) %>% gsub("\\.\\*", "_", .)

```
## Quantile normalise log2(RPKM + 1)

```{r aggregated_counts_fpkm_log2}

aggregated.peak.fpkm <-
  rpkm(aggregated.peak.counts,
       log = F,
       gene.length = peak.counts$Length)

# Keep peaks with more than 10 FPKMs in any condition.

aggregated.keep <- rowSums(aggregated.peak.fpkm > 10) >= 1

aggregated.peak.fpkm <- aggregated.peak.fpkm[aggregated.keep, ]
aggregated.peak.fpkm <- log2(aggregated.peak.fpkm + 1)
aggregated.peak.fpkm <- normalizeQuantiles(aggregated.peak.fpkm)

```

## Discard low coefficient of variation peaks.

```{r cv_20_peaks}
cv.20 <- apply(aggregated.peak.fpkm, 1, function(x) cv(x) * 100) > 20
cv.20 %>% table()

aggregated.peak.fpkm.high.cv <- aggregated.peak.fpkm[cv.20, ]
```

## Run fuzzy clustering.
```{r fuzzy_clustering}

exprSet <- ExpressionSet(assayData = aggregated.peak.fpkm.high.cv)

s.exprSet <- standardise(exprSet)

m <- mestimate(s.exprSet)

set.seed(1234)
cl.s.exprSet <- mfuzz(s.exprSet, c = 8, m = m)

# Retrieve peak cluster membership

peak.cluster.membership <- as.data.frame(cl.s.exprSet$cluster) %>% 
  dplyr::mutate(PeakID = rownames(.))
colnames(peak.cluster.membership)[1] <- "Cluster"

```
### Filter high affinity peaks.

```{r core_peaks}

core.exprSet <- acore(s.exprSet,
                      cl = cl.s.exprSet,
                      min.acore = 0.8)

# Peak ids of high affinity peaks
core.exprSet.peaks.id <- sapply(c(1:8), function(x) as.vector(core.exprSet[[x]]$NAME)
)
core.exprSet.peaks.id <- as.character(unlist(core.exprSet.peaks.id))

```

## Create a table of high cluster affinity peaks.
It contains z-scores and peak details.

```{r}

# Peak z-scores

peak.z.scores <- as.data.frame(s.exprSet@assayData$exprs) %>% 
  dplyr::mutate(., PeakID = rownames(.))

# Table of high affinity peaks annotated.
high.affinity.peaks.annotated <- dplyr::left_join(peak.z.scores,
                                                  peak.cluster.membership,
                                                  by = "PeakID") %>%
  dplyr::left_join(peak.counts[, 1:6], by = "PeakID") %>%
  dplyr::filter(PeakID %in% core.exprSet.peaks.id)

```

# Plots cluster dynamics (mean +/- SD).
Please note that for the benefit of biological flow clusters have been relabeled as follows and has been indicated as Cluster_Labels accross the manuscript, source data and supplementary tables.
1 = "C 1", 2 = "C 2", 3 = "C 7", 4 = "C 4", 5 = "C 3", 6 = "C 6", 7 = "C 5", 8 = "C 8".

## High affinity cluster membership counts
```{r}

core.exprSet.cluster.counts <- lapply(core.exprSet, nrow) %>%
  unlist() %>% as.data.frame() %>% 
  mutate(Cluster = c(1:8))

colnames(core.exprSet.cluster.counts)[1] <- "members"
```

