---
title: | 
  | Script for analysing bulk ATAC-seq of reprogramming intermediate
author: |
  | Fernando J. Rossello
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    toc: true
    toc_depth: 2
    toc_float: 
      collapsed: false
fontsize: 12pt
pagetitle: "bkATAC"
---

# Preamble
## Things to note
- All input files can be downloaded from http://hrpi.ddnetbio.com/
- All input files are assumed to be in the data folder.
- For reproducibility purposes, the randomisation seed used throughout the manuscript was saved as an object and it should be loaded in all relevant sections (object `seed` in `data/seed.Rdata`loaded in each section's code).
- All genome browser tracks can be explored at http://hrpi.ddnetbio.com at the _bulk ATAC-seq iPSC reprog. section_.

## Load required libraries.

```{r session_setup, include = TRUE, echo=TRUE, results='hide', message=FALSE, warning=FALSE}
library(dplyr)
library(edgeR)
library(irlba)
library(sjstats)
library(Mfuzz)
library(patchwork)
library(data.table)
library(ggplot2)
library(ggrepel)
library(knitr)
library(annotatr)
# library(GenomicRanges)
# library(GenomicFeatures)
# rm(list=ls())
```
## Define colour palettes.

```{r color_labels}

colorMedia = c("black","darkorange","blue","darkolivegreen2","red2","pink2")
names(colorMedia) = c("D0, 3, 7","Primed","t2iLGoY","5iLAF","NHSM","RSeT")
shapeTime = c(0,1,2,15,16,17,18)
names(shapeTime) = c("D0","D3","D7","D13","D21","P3","P10")

gplot.theme <- theme(panel.border=element_blank(), 
        panel.background=element_blank(),
        axis.line=element_line(),
        axis.line.y=element_line("black", size=1),
        axis.line.x=element_line("black", size=1),
        axis.text=(element_text(colour="black", size = 6, face="bold")),
        axis.title=(element_text(colour="black", size = 8, face="bold")),
        axis.ticks = (element_line("black", size=0.5)),
        legend.title = element_text(colour="black", size = 9, face="bold"),
        legend.text = element_text(colour="black", size= 8, face="bold"),
        legend.key.size=unit(0.1,"cm"),
        # legend.margin = unit(0, "cm"),
        legend.position="none",
        plot.margin=unit(c(0,2,0,0),"mm"),
        strip.text.x = element_text(size = 6, colour = "black", face = "bold", margin = margin(0,0,0,0, "cm"))
  )
```

# Exploratory Analysis - PCA.
## Import read counts of consensus peak set.
```{r import_read_counts}

peak.counts <-
  read.table(file = "../data/bkATAC_consensus_peak_set_counts.txt",
             header = T,
             stringsAsFactors = F)
```

### Reorder samples and save peak information.
```{r reorder_samples}
order <- colnames(peak.counts)[7:ncol(peak.counts)]
order <- order[c(1, 2, 11:14, 3, 5, 7, 9, 19, 21, 15, 17, 4, 6, 8, 10, 20, 22, 16, 18)]
order <- factor(order, levels = order)

peak.info <- colnames(peak.counts)[1:6]
```

### Reorder peak counts table.

```{r prepare_counts_table}

peak.counts <- peak.counts[, c(peak.info, levels(order))]

# Change Geneid column name to PeakID

colnames(peak.counts)[1] <- "PeakID"
rownames(peak.counts) <- peak.counts$PeakID
```

## Within and between library normalization.

### Quantile normalise log2(RPKM + 1)

```{r quantile_log2}

peaks.fpkm <-
  rpkm(peak.counts[, 7:ncol(peak.counts)],
       log = F, gene.length = peak.counts$Length)

# Keep peaks with more than 5 FPKMs in at least 2 samples for exploratory analysis.

keep <- rowSums(peaks.fpkm > 5) >= 2

peaks.fpkm <- peaks.fpkm[keep, ]
peaks.fpkm <- log2(peaks.fpkm + 1)
peaks.fpkm <- normalizeQuantiles(peaks.fpkm)
```

## PCA 

```{r pca}

# Prepare annotation table.
pca.plot.annotation <- data.table(sample_id = colnames(peaks.fpkm))

pca.plot.annotation$sample_labels <- c(
  paste(rep("FM_", 3), rep("D", 3), rep(c(0, 3, 7), each = 2), rep(c("_R1", "_R2")), sep = ""),
  paste(rep("PM_", 2), rep("D", 2), rep(c(13, 21), each = 2), rep(c("_R1", "_R2")), sep = ""),
  paste(rep("PM_", 2), rep("P", 2), rep(c(3, 10), each = 2), rep(c("_R1", "_R2")), sep = ""),
  paste(rep("NM_", 2), rep("D", 2), rep(c(13, 21), each = 2), rep(c("_R1", "_R2")), sep = ""),
  paste(rep("NM_", 2), rep("P", 2), rep(c(3, 10), each = 2), rep(c("_R1", "_R2")), sep = "")
)

pca.plot.annotation$media = "D0, 3, 7"
pca.plot.annotation[grep("PM", sample_labels)]$media = "Primed"
pca.plot.annotation[grep("NM", sample_labels)]$media = "t2iLGoY"
pca.plot.annotation$media = factor(pca.plot.annotation$media, 
                           levels = c("D0, 3, 7", "Primed", "t2iLGoY"))
pca.plot.annotation$timept = tstrsplit(pca.plot.annotation$sample_labels, "_")[[2]]
pca.plot.annotation$timept = factor(pca.plot.annotation$timept, 
                            levels = unique(pca.plot.annotation$timept))

base::load("../data/seed.Rdata")
.Random.seed <- seed
pca <- prcomp_irlba(t(peaks.fpkm))

pca.plot.annotation$PC1 = pca$x[,1]
pca.plot.annotation$PC2 = pca$x[,2]
pca.plot.annotation$PC3 = pca$x[,3]

ggplot(pca.plot.annotation, aes(PC1, PC2, color = media, label = timept)) +
  geom_point(size = 3) + geom_text_repel(size = 5) +
  scale_color_manual(values = colorMedia[1:3]) +
  scale_shape_manual(values = c(16, 15)) + theme_classic(base_size = 24)
  
ggplot(pca.plot.annotation, aes(PC1, PC3, color = media, label = timept)) +
  geom_point(size = 3) + geom_text_repel(size = 5) +
  scale_color_manual(values = colorMedia[1:3]) +
  scale_shape_manual(values = c(16, 15)) + theme_classic(base_size = 24)


```

# Fuzzy clustering.

## Aggregate sample counts means and calculate RPKM.

```{r aggregate_sample_counts_rpkm}
# Prepare conditions labels.

conditions.labels <-
  c(
  "D0",
  "D3",
  "D7",
  "D13.*Primed",
  "D21.*Primed",
  "P3.*Primed",
  "P10.*Primed",
  "D13.*Smith",
  "D21.*Smith",
  "P3.*Smith",
  "P10.*Smith"
  )

# Aggregate peak counts.
aggregated.peak.counts <-
  sapply(conditions.labels, function(x)
    rowSums(peak.counts[, grep(x,
                               colnames(peak.counts),
                               value = TRUE)]))

colnames(aggregated.peak.counts) <-
  colnames(aggregated.peak.counts) %>% gsub("\\.\\*", "_", .)

```
## Quantile normalise log2(RPKM + 1)

```{r aggregated_counts_fpkm_log2}

aggregated.peak.fpkm <-
  rpkm(aggregated.peak.counts,
       log = F,
       gene.length = peak.counts$Length)

# Keep peaks with more than 10 FPKMs in any condition.

aggregated.keep <- rowSums(aggregated.peak.fpkm > 10) >= 1

aggregated.peak.fpkm <- aggregated.peak.fpkm[aggregated.keep, ]
aggregated.peak.fpkm <- log2(aggregated.peak.fpkm + 1)
aggregated.peak.fpkm <- normalizeQuantiles(aggregated.peak.fpkm)

```

## Discard low coefficient of variation peaks.

```{r cv_20_peaks}
cv.20 <- apply(aggregated.peak.fpkm, 1, function(x) cv(x) * 100) > 20
cv.20 %>% table()

aggregated.peak.fpkm.high.cv <- aggregated.peak.fpkm[cv.20, ]

```

## Run fuzzy clustering.

```{r fuzzy_clustering}

exprSet <- ExpressionSet(assayData = aggregated.peak.fpkm.high.cv)

s.exprSet <- standardise(exprSet)

m <- mestimate(s.exprSet)
# Create seed as save it as an object.
# set.seed(1234)
# seed <- .Random.seed
# save(seed, file = "../data/seed.Rdata")

load("../data/seed.Rdata")
.Random.seed <- seed

cl.s.exprSet <- mfuzz(s.exprSet, c = 8, m = m)


# Retrieve peak cluster membership

peak.cluster.membership <- as.data.frame(cl.s.exprSet$cluster) %>% 
  dplyr::mutate(PeakID = rownames(.))
colnames(peak.cluster.membership)[1] <- "Cluster"

```
### Filter high affinity peaks.

```{r core_peaks}

core.exprSet <- acore(s.exprSet,
                      cl = cl.s.exprSet,
                      min.acore = 0.8)

# Peak ids of high affinity peaks
core.exprSet.peaks.id <- sapply(c(1:8), function(x) as.vector(core.exprSet[[x]]$NAME)
)
core.exprSet.peaks.id <- as.character(unlist(core.exprSet.peaks.id))

```

## Create a table of high cluster affinity peaks.
It contains z-scores and peak details.
Please note that for the benefit of biological flow, clusters have been relabeled as follows: 1 = "C 1", 2 = "C 2", 3 = "C 7", 4 = "C 4", 5 = "C 3", 6 = "C 6", 7 = "C 5", 8 = "C 8". This has also been indicated as `Cluster_Labels` accross the manuscript's source data and supplementary tables.
.

```{r high_affinity_peaks_z-s}

# Peak z-scores

peak.z.scores <- as.data.frame(s.exprSet@assayData$exprs) %>% 
  dplyr::mutate(., PeakID = rownames(.))

# Table of high affinity peaks annotated.
high.affinity.peaks.annotated <- dplyr::left_join(peak.z.scores,
                                                  peak.cluster.membership,
                                                  by = "PeakID") %>%
  dplyr::left_join(peak.counts[, 1:6], by = "PeakID") %>%
  dplyr::filter(PeakID %in% core.exprSet.peaks.id)

high.affinity.peaks.annotated <-
  high.affinity.peaks.annotated %>% mutate(
  Cluster_Label = case_when(
  Cluster == "1" ~ "C 1",
  Cluster == "2" ~ "C 2",
  Cluster == "3" ~ "C 7",
  Cluster == "4" ~ "C 4",
  Cluster == "5" ~ "C 3",
  Cluster == "6" ~ "C 6",
  Cluster == "7" ~ "C 5",
  TRUE ~ "C 8"
  )
  )
  
```

## Cluster dynamics (mean +/- SD).

### High affinity cluster membership counts table.
```{r membership_counts}

core.exprSet.cluster.counts <- lapply(core.exprSet, nrow) %>%
  unlist() %>% as.data.frame() %>% 
  mutate(Cluster = c(1:8))

colnames(core.exprSet.cluster.counts)[1] <- "members"

core.exprSet.cluster.counts <-
  core.exprSet.cluster.counts %>% mutate(
  Cluster_Label = case_when(
  Cluster == "1" ~ "C 1",
  Cluster == "2" ~ "C 2",
  Cluster == "3" ~ "C 7",
  Cluster == "4" ~ "C 4",
  Cluster == "5" ~ "C 3",
  Cluster == "6" ~ "C 6",
  Cluster == "7" ~ "C 5",
  TRUE ~ "C 8"
  )
  )
  
core.exprSet.cluster.counts$Cluster_Label <- factor(x = core.exprSet.cluster.counts$Cluster_Label,
                                                    levels = paste("C", 1:8, sep = " "))
knitr::kable(core.exprSet.cluster.counts %>% select(Cluster_Label, members) %>% arrange(Cluster_Label))
```

## Cluster dynamics plots.

### Table.

```{r  cluster_dynamics_table}
high.affinity.peaks.melt <-
  melt(select(
  high.affinity.peaks.annotated,
  -c("Chr", "Start", "End", "Strand", "Length", "Cluster_Label")
  ),
  id = c("PeakID", "Cluster"))

# Define stage and media

high.affinity.peaks.melt <-
  high.affinity.peaks.melt %>% mutate(stage = case_when(
  grepl("D13", variable) ~ "D13",
  grepl("D21", variable) ~ "D21",
  grepl("P3", variable) ~ "P3",
  grepl("P10", variable) ~ "P10",
  TRUE ~ as.character(variable)
  )) %>%
  mutate(media = case_when(
  grepl("D0|D3|D7", variable) ~ "FM",
  grepl("Smith", variable) ~ "NM",
  grepl("Primed", variable) ~ "PM"
  ))
  
# Order Stages                                                          
high.affinity.peaks.melt$stage <-
  factor(high.affinity.peaks.melt$stage,
  levels = c("D0", "D3", "D7", "D13", "D21", "P3", "P10"))
  
# Order Clusters
high.affinity.peaks.melt$Cluster <-
  factor(high.affinity.peaks.melt$Cluster, levels = c(1, 2, 5, 4, 7, 6, 3, 8))



cluster.dynamics.plot.values <-
  high.affinity.peaks.melt %>% group_by(media, stage, Cluster) %>% summarise(
  mean = mean(value),
  sd = sd(value)
  )
  
# Define Cluster labels (please see comment above)

cluster.dynamics.plot.values <-
  cluster.dynamics.plot.values %>% mutate(
  Cluster_Label = case_when(
  Cluster == "1" ~ "C 1",
  Cluster == "2" ~ "C 2",
  Cluster == "3" ~ "C 7",
  Cluster == "4" ~ "C 4",
  Cluster == "5" ~ "C 3",
  Cluster == "6" ~ "C 6",
  Cluster == "7" ~ "C 5",
  TRUE ~ "C 8"
  )
  )

cluster.dynamics.plot.values$Cluster_Label  <-
  factor(
  x = cluster.dynamics.plot.values$Cluster_Label,
  levels = c("C 1", "C 2", "C 3", "C 4", "C 5", "C 6", "C 7", "C 8")
  )


```
### Cluster dynamics plot.
Ribbon equals mean +/- SD.
```{r cluster_dynamics_plot, fig.align="center", fig.height=9, fig.width=4.5}

ggplot(data = NULL, aes(stage, mean, group = 1)) +
  geom_line(
  data = dplyr::filter(cluster.dynamics.plot.values, media %in% c("FM", "PM")),
  color = "#FF8C00",
  size = 0.5
  ) +
  geom_line(
  data = dplyr::filter(
  cluster.dynamics.plot.values,
  media %in% c("FM", "PM"),
  stage %in% c("D0", "D3", "D7")
  ),
  color = "black",
  size = 0.5
  ) +
  geom_ribbon(
  data = dplyr::filter(
  cluster.dynamics.plot.values,
  media %in% c("FM", "PM"),
  stage %in% c("D0", "D3", "D7")
  ),
  aes(
  x = stage,
  ymax = mean + sd,
  ymin = mean - sd
  ),
  fill = "black",
  alpha = 0.5
  ) +
  geom_ribbon(
  data = dplyr::filter(
  cluster.dynamics.plot.values,
  media %in% c("FM", "PM"),
  stage %in% c("D7", "D13", "D21", "P3", "P10")
  ),
  aes(
  x = stage,
  ymax = mean + sd,
  ymin = mean - sd
  ),
  fill = "#FF8C00",
  alpha = 0.5
  ) +
  geom_line(
  data = dplyr::filter(cluster.dynamics.plot.values, media %in% c("FM", "NM")),
  color = "#2302FE",
  size = 0.5
  ) +
  geom_line(
  data = dplyr::filter(
  cluster.dynamics.plot.values,
  media %in% c("FM", "NM"),
  stage %in% c("D7", "D13", "D21", "P3", "P10")
  ),
  color = "#2302FE",
  size = 0.5
  ) +
  geom_line(
  data = dplyr::filter(
  cluster.dynamics.plot.values,
  media %in% c("FM", "NM"),
  stage %in% c("D0", "D3", "D7")
  ),
  color = "black",
  size = 0.5
  ) +
  geom_ribbon(
  data = dplyr::filter(
  cluster.dynamics.plot.values,
  media %in% c("FM", "NM"),
  stage %in% c("D0", "D3", "D7")
  ),
  aes(
  x = stage,
  ymax = mean + sd,
  ymin = mean - sd
  ),
  fill = "black",
  alpha = 0.5
  ) +
  geom_ribbon(
  data = dplyr::filter(
  cluster.dynamics.plot.values,
  media %in% c("FM", "NM"),
  stage %in% c("D7", "D13", "D21", "P3", "P10")
  ),
  aes(
  x = stage,
  ymax = mean + sd,
  ymin = mean - sd
  ),
  fill = "#2302FE",
  alpha = 0.5
  ) +
  facet_wrap( ~ Cluster_Label, nrow = 4, ncol = 2) +
  gplot.theme + theme(legend.position = "none", axis.title = element_blank())

```

# Basic cluster Annotation (hg19).

```{r basic_hg19_annotation, message=FALSE, warning=FALSE}

hg19.basic.genes <- c('hg19_basicgenes', 'hg19_genes_intergenic')

hg19.basic.genes.annotation <- build_annotations(genome = "hg19", annotations = hg19.basic.genes)

high.affinity.peaks.annotated.granges <-
  makeGRangesFromDataFrame(high.affinity.peaks.annotated,
  keep.extra.columns = T)
  
high.affinity.peaks.annotated.granges <-
  annotate_regions(
  regions = high.affinity.peaks.annotated.granges,
  annotations = hg19.basic.genes.annotation,
  ignore.strand = T,
  quiet = FALSE
  )

```

## Basic annotation table.

```{r}

# Annotation type data frame

high.affinity.peaks.annotated.granges.df <-
  data.frame(high.affinity.peaks.annotated.granges) %>%
  dplyr::select(PeakID, Cluster_Label, annot.type) %>%
  dplyr::distinct(PeakID, annot.type, .keep_all = T)#


basic.annotation.type.freqs <-
  high.affinity.peaks.annotated.granges.df %>%
  group_by(Cluster_Label, annot.type) %>%
  summarise(n = n()) %>%
  mutate(total = sum(n), rel.freq = n / total)


basic.annotation.type.freqs$Cluster_Label <- factor(x = basic.annotation.type.freqs$Cluster_Label,
                                                    levels = paste("C", 1:8, sep = " "))
knitr::kable(
  basic.annotation.type.freqs %>% dplyr::select(Cluster_Label, annot.type, rel.freq) %>% arrange(Cluster_Label)
  )

```

## Basic annotation plot.

```{r basic_annotation_plot}

ggplot(data = basic.annotation.type.freqs, aes(x = Cluster_Label, y = rel.freq, fill = annot.type)) +
  geom_bar(position = "stack", stat = "identity") +
  coord_flip() +
  scale_x_discrete(limits = rev(levels(
  basic.annotation.type.freqs$Cluster_Label
  ))) +
  scale_fill_brewer(palette = "Paired") +
  xlab("ATAC-seq clusters") +
  ylab("Proportion of Cluster specific ATAC-seq peaks class") +
  labs(fill = "Class") +
  gplot.theme + theme(
  legend.position = "right",
  legend.title = element_text(size = 10),
  legend.text = element_text(size = 9),
  axis.title = element_text(size = 10),
  axis.text = element_text(size = 9)
  )
```


# Integration of chromatin accessibility and expression profiles.

Profile of genes associated with primed and naive reprogramming and somatic regulatory elements. 
We also show the global gene expression profiles of genes associated with peaks in the clusters identified above.

Based on comprehensively annotated clusters (annotation with Homer, see methods) and peaks asociated with gene TSSs (see methods).

For simplicity, objects with the annotated clusters and z-scores of expressed genes are loaded `data/bkATAC_annotated_clusters.Rdata` and `data/bkATAC_gene_expression_bulk_RNA_seq.Rdata`)

## Load annotated clusters and process.
```{r annotated_clusters}
# Load annotated cluster and prepare peaks closest to TSSs


load(file = "../data/bkATAC_annotated_clusters.Rdata") # Creates `annotated.clusters` object.

# Filter peaks with no Ensembl ID associated.

annotated.clusters.table <- ldply(annotated.clusters, rbind) %>% filter(!Nearest.Ensembl == "" ) %>% .[, -1] %>% dplyr::select(-c("Peak.Score", "Focus.Ratio.Region.Size"))

# In cases were a genes has more than one associated peak, the closest peak to the TSS is selected.
annotated.clusters.table.nearest.TSS <-
  as.data.frame(annotated.clusters.table) %>%
  dplyr::group_by(Nearest.Ensembl) %>%
  dplyr::slice(which.min(abs(Distance.to.TSS)))

# Add cluster information and chromatin accessiblity z-scores.  
annotated.clusters.table.nearest.TSS <-
  dplyr::left_join(
  annotated.clusters.table.nearest.TSS,
  dplyr::select(
  high.affinity.peaks.annotated,
  -c("Chr", "Start", "End", "Strand", "Length")
  ),
  by = "PeakID"
  )

```

## Peak associated genes.

### Expression table.
```{r z_scores_rna_seq}
base::load(file = "../data/bkATAC_gene_z_scores_bkRNA_seq.Rdata") # Load RNA-seq gene z-scores.

# Filter out genes with no associated cluster and add cluster information

s.exprSet.mean.log2.rpkm.bulk.RNA.seq.filtered.low.count.genes.atac.seq.changing.cluster.tss <-  dplyr::inner_join(s.exprSet.mean.log2.rpkm.bulk.RNA.seq.filtered.low.count.genes,
                  homer.annotated.changing.clusters.details.table.nearest.TSS.cluster.ids, by = "Ensembl_id")

```

## Plots
```{r}
##### Peaks closest to TSS - All detected genes. 


s.exprSet.mean.log2.rpkm.bulk.RNA.seq.filtered.low.count.genes.atac.seq.changing.cluster.tss

gene.ids <- bulk.RNA.seq.raw.counts[, 1:2]
colnames(gene.ids)[1] <- "Ensembl_id"

s.exprSet.mean.log2.rpkm.bulk.RNA.seq.filtered.low.count.genes.atac.seq.changing.cluster.tss.annotated <-
  dplyr::inner_join(
  s.exprSet.mean.log2.rpkm.bulk.RNA.seq.filtered.low.count.genes.atac.seq.changing.cluster.tss,
  gene.ids,
  by = "Ensembl_id"
  )
for (i in c(1:8)){
  write.table(s.exprSet.mean.log2.rpkm.bulk.RNA.seq.filtered.low.count.genes.atac.seq.changing.cluster.tss.annotated %>% dplyr::filter(Cluster == i) %>% dplyr::select(c("Ensembl_id", "Cluster", "geneName")),
  file = paste("~/projects/Polo_Group/Ethan_Liu/Ethan_Liu_ATAC-seq_Human_Reprogramming_human_samples/analysis/R/data/fuzzy_clustering/combined_sets/cluster_analysis/bulk_RNA-seq_genes_over_ATAC-seq_changing_clusters_peaks_TSS/",
  "cluster_",
  i,
  "_gene_ids.csv",sep = ""),
  row.names = F,
  sep = "\t",
  quote = F)
}


s.exprSet.mean.log2.rpkm.bulk.RNA.seq.filtered.low.count.genes.atac.seq.changing.cluster.tss.melt <-
  melt(
  s.exprSet.mean.log2.rpkm.bulk.RNA.seq.filtered.low.count.genes.atac.seq.changing.cluster.tss,
  id = c("Ensembl_id", "Cluster")
  )
  
s.exprSet.mean.log2.rpkm.bulk.RNA.seq.filtered.low.count.genes.atac.seq.changing.cluster.tss.melt <-
  s.exprSet.mean.log2.rpkm.bulk.RNA.seq.filtered.low.count.genes.atac.seq.changing.cluster.tss.melt %>%
  mutate(media = ifelse(grepl("Fibroblast", variable), "FM",
  ifelse(grepl("Primed", variable), "PM", "NM")))  %>%
  mutate(stage = gsub(".*_", "", variable))

s.exprSet.mean.log2.rpkm.bulk.RNA.seq.filtered.low.count.genes.atac.seq.changing.cluster.tss.melt$stage <-
  factor(
  s.exprSet.mean.log2.rpkm.bulk.RNA.seq.filtered.low.count.genes.atac.seq.changing.cluster.tss.melt$stage,
  levels = c("D00", "D03", "D07", "D13", "D21", "P03", "P10"),
  labels = c("D0", "D3", "D7", "D13", "D21", "P3", "P10")
  )
  
s.exprSet.mean.log2.rpkm.bulk.RNA.seq.filtered.low.count.genes.atac.seq.changing.cluster.tss.counts <-
  s.exprSet.mean.log2.rpkm.bulk.RNA.seq.filtered.low.count.genes.atac.seq.changing.cluster.tss %>%
  group_by(Cluster) %>%
  dplyr::count()



s.exprSet.mean.log2.rpkm.bulk.RNA.seq.filtered.low.count.genes.atac.seq.changing.cluster.tss.melt$Cluster <- factor(s.exprSet.mean.log2.rpkm.bulk.RNA.seq.filtered.low.count.genes.atac.seq.changing.cluster.tss.melt$Cluster, levels = c(1, 2, 5, 4, 7, 6, 3, 8))
cluster.labels.rna.seq <- c("1" = "C 1 (n = 1823)", "2" = "C 2 (n = 1347)", "3" = "C 7 (n = 790)", "4" = "C 4 (n = 339)", "5" = "C 3 (n = 892)", "6" = "C 6 (n = 1027)", "7" = "C 5 (n = 1239)", "8" = "C 8 (n = 1547)")
atac.seq.clustes.tss.peaks.bulk.RNA.seq.all.genes.ggplot <-
  ggplot(data = s.exprSet.mean.log2.rpkm.bulk.RNA.seq.filtered.low.count.genes.atac.seq.changing.cluster.tss.melt %>%
  dplyr::filter(media %in% c("FM", "PM")),
  aes(stage, value, group = Ensembl_id, colour = media)) +
  stat_summary(aes(group = 1),
  fun.y = mean,
  geom = "line",
  size = 1) +
  stat_summary(
  data = s.exprSet.mean.log2.rpkm.bulk.RNA.seq.filtered.low.count.genes.atac.seq.changing.cluster.tss.melt %>%
  dplyr::filter(media %in% c("FM", "NM")),
  aes(group = 1),
  fun.y = mean,
  geom = "line",
  size = 1
  ) +
  scale_color_manual(values = c("FM" = "black", "NM" = "#0000FF", "PM" = "#FF8C00"),
  name = "Culture Media") +
  facet_wrap(~ Cluster, nrow = 1, ncol = 8) +
  # facet_wrap(~ Cluster, nrow = 1, ncol = 8, labeller = labeller(Cluster = cluster.labels.rna.seq)) +
  ylab("Gene expression (z-score)") +
  xlab("Reprogramming stages") +
  # ggtitle("Fuzzy clustering")  +
  # geom_text(
  # data = s.exprSet.mean.log2.rpkm.bulk.RNA.seq.filtered.low.count.genes.atac.seq.changing.cluster.tss.counts,
  # aes(
  # x = 2,
  # y = -0.4,
  # label = paste("n = ", n)
  # ),
  # colour = "black",
  # inherit.aes = FALSE,
  # parse = FALSE,
  # size = 2.5,
  # fontface = "bold"
  # ) +
  gplot.theme
print(atac.seq.clustes.tss.peaks.bulk.RNA.seq.all.genes.ggplot)
ggsave(
  plot = atac.seq.clustes.tss.peaks.bulk.RNA.seq.all.genes.ggplot,
  filename = "~/projects/Polo_Group/Ethan_Liu/Ethan_Liu_ATAC-seq_Human_Reprogramming_human_samples/analysis/R/plots/Figure_PDFs/Figure_4_bulk_RNA-seq_over_ATAC-seq_changing_clusters_peaks_tss_resized_reordered_horizontal_no_labels.pdf",
  height = 5,
  width = 28,
  units = "cm",
  dpi = 200,
  scale = 1,
  device = "pdf"
  )

extended.figure.6.c.source.data <- select(
  s.exprSet.mean.log2.rpkm.bulk.RNA.seq.filtered.low.count.genes.atac.seq.changing.cluster.tss.melt,
  -variable
  )

extended.figure.6.c.source.data <- extended.figure.6.c.source.data %>% mutate(
  Cluster_Label = case_when(
  Cluster == "1" ~ "C 1",
  Cluster == "2" ~ "C 2",
  Cluster == "3" ~ "C 7",
  Cluster == "4" ~ "C 4",
  Cluster == "5" ~ "C 3",
  Cluster == "6" ~ "C 6",
  Cluster == "7" ~ "C 5",
  TRUE ~ "C 8"
  )
  )
extended.figure.6.c.source.data$Cluster_Label  <-
  factor(x = extended.figure.6.c.source.data$Cluster_Label,
  levels = c("C 1", "C 2", "C 3", "C 4", "C 5", "C 6", "C 7", "C 8"))

# Plot test

ggplot(data = extended.figure.6.c.source.data %>%
  dplyr::filter(media %in% c("FM", "PM")),
  aes(stage, value, group = Ensembl_id, colour = media)) +
  stat_summary(aes(group = 1),
  fun.y = mean,
  geom = "line",
  size = 1) +
  stat_summary(
  data = extended.figure.6.c.source.data %>%
  dplyr::filter(media %in% c("FM", "NM")),
  aes(group = 1),
  fun.y = mean,
  geom = "line",
  size = 1
  ) +
  scale_color_manual(values = c("FM" = "black", "NM" = "#0000FF", "PM" = "#FF8C00"),
  name = "Culture Media") +
  facet_wrap(~ Cluster_Label, nrow = 1, ncol = 8) +
  ylab("Gene expression (z-score)") +
  xlab("Reprogramming stages") +
  gplot.theme

write_excel_csv(x = select(extended.figure.6.c.source.data,-Cluster),
                path = "~/projects/Polo_Group/Ethan_Liu/Ethan_Liu_ATAC-seq_Human_Reprogramming_human_samples/analysis/R/data/Source_Data/extended.figure.6.c.source.data.csv")
```

##### ATAC-seq peaks dynamics closest to TSS selected genes
```{r atac-seq peaks dynamics closest to tss seleceted genes}

homer.annotated.changing.clusters.details.table.nearest.TSS.by.gene <- homer.annotated.changing.clusters.details.table.nearest.TSS %>% ungroup %>% dplyr::select("Gene.Name", "D0", "D3", "D7", "D13_Primed", "D21_Primed", "P3_Primed", "P10_Primed", "D13_Smith", "D21_Smith", "P3_Smith", "P10_Smith")

homer.annotated.changing.clusters.details.table.nearest.TSS.by.gene.melt <- melt(homer.annotated.changing.clusters.details.table.nearest.TSS.by.gene, id = c("Gene.Name"))
### ADD MUTATE IFELSE FOR MEDIA AND STAGE POST MELT

homer.annotated.changing.clusters.details.table.nearest.TSS.by.gene.melt <- homer.annotated.changing.clusters.details.table.nearest.TSS.by.gene.melt %>% mutate(stage = ifelse(variable == "D0", "D0",ifelse(variable == "D3", "D3", ifelse(variable == "D7", "D7", ifelse(variable == "D13_Primed" | variable == "D13_Smith", "D13", ifelse(variable == "D21_Primed" | variable == "D21_Smith", "D21", ifelse(variable == "P3_Primed" | variable == "P3_Smith", "P3", "P10")))))))

homer.annotated.changing.clusters.details.table.nearest.TSS.by.gene.melt <- homer.annotated.changing.clusters.details.table.nearest.TSS.by.gene.melt %>% 
  mutate(media = case_when(grepl("D0|D3|D7", variable) ~ "FM",
                           grepl("Smith", variable) ~ "NM",
                           grepl("Primed", variable) ~ "PM"))
                                                             
homer.annotated.changing.clusters.details.table.nearest.TSS.by.gene.melt$stage <-
  factor(
  homer.annotated.changing.clusters.details.table.nearest.TSS.by.gene.melt$stage,
  levels = c("D0", "D3", "D7", "D13", "D21", "P3", "P10")
  )

homer.annotated.changing.clusters.details.table.nearest.TSS.by.gene.melt.candidates <- dplyr::filter(homer.annotated.changing.clusters.details.table.nearest.TSS.by.gene.melt, Gene.Name %in% levels(
  z.score.figure.4.bulk.RNA.seq.candidates.canonical.ensembl$geneName))

# homer.annotated.changing.clusters.details.table.nearest.TSS.by.gene.melt.candidates$Gene.Name <- factor(homer.annotated.changing.clusters.details.table.nearest.TSS.by.gene.melt.candidates$Gene.Name, levels = levels(
#   z.score.figure.4.bulk.RNA.seq.candidates.canonical.ensembl$geneName))

homer.annotated.changing.clusters.details.table.nearest.TSS.by.gene.melt.candidates$Gene.Name <- factor(homer.annotated.changing.clusters.details.table.nearest.TSS.by.gene.melt.candidates$Gene.Name, levels = c("SNAI1", "TWIST2", "ZEB2", "BHLHE40", "F11R", "ZIC5", "OCLN", "NLRP7"))


rna.seq.over.atac.seq.dynamics <- ggplot(
  data = homer.annotated.changing.clusters.details.table.nearest.TSS.by.gene.melt.candidates %>%
  filter(
  media %in% c("FM", "PM")),
  aes(stage, value, group = Gene.Name, colour = media)
  ) +
  geom_line(size = 1.5) +
  geom_line(
  data = homer.annotated.changing.clusters.details.table.nearest.TSS.by.gene.melt.candidates %>%
  filter(
  media %in% c("FM", "NM")
  ),
  aes(stage, value, group = Gene.Name, colour = media), size = 1.5
  ) +
  scale_color_manual(values = c("FM" = "black", "NM" = "#0000FF", "PM" = "#FF8C00"),
  name = "Culture Media") +
  facet_wrap( ~ Gene.Name, nrow = 8, ncol = 1) +
  ylab("Chromatin Accessiblity (z-score)") +
  xlab("Reprogramming stages") +
  gplot.theme
  ggsave(plot = rna.seq.over.atac.seq.dynamics,
  "~/projects/Polo_Group/Ethan_Liu/Ethan_Liu_ATAC-seq_Human_Reprogramming_human_samples/analysis/R/plots/Figure_PDFs/Figure_4_bulk_RNA-seq_over_ATAC-seq_changing_clusters_line_plots_closest_to_tss_dynamics_reordered.pdf",
  height = 15,
  width = 4.5,
  units = "cm",
  dpi = 300,
  scale = 1
  )
```
##### Selected gene expression plots - Cluster basd
```{r}
figure.4.bulk.RNA.seq.candidates.canonical.ensembl <- queryMany(figure.4.bulk.RNA.seq.candidates.canonical$hg19.kgXref.geneSymbol, fields = "ensembl.gene", scope = "symbol", species = "human")

z.score.figure.4.bulk.RNA.seq.candidates.canonical.ensembl <-
  filter(
    s.exprSet.mean.log2.rpkm.bulk.RNA.seq.filtered.low.count.genes.atac.seq.changing.cluster.tss.melt,
    Ensembl_id %in% c(
      figure.4.bulk.RNA.seq.candidates.canonical.ensembl$ensembl %>% unlist()
    )
  )
z.score.figure.4.bulk.RNA.seq.candidates.canonical.ensembl <- inner_join(z.score.figure.4.bulk.RNA.seq.candidates.canonical.ensembl, gene.ids, by = "Ensembl_id")
z.score.figure.4.bulk.RNA.seq.candidates.canonical.ensembl$Cluster <- factor(z.score.figure.4.bulk.RNA.seq.candidates.canonical.ensembl$Cluster, levels = c(1:8))

z.score.figure.4.bulk.RNA.seq.candidates.canonical.ensembl$geneName <- factor(z.score.figure.4.bulk.RNA.seq.candidates.canonical.ensembl$geneName,
                                                                              levels = c("SNAI1", "TWIST2", "ZEB2", "BHLHE40", "F11R", "ZIC5", "OCLN", "NLRP7"))
selected.genes.line.plots <- ggplot(data = z.score.figure.4.bulk.RNA.seq.candidates.canonical.ensembl %>% 
         filter(media %in% c("FM", "PM")), 
       aes(stage, value, group = geneName, colour = media)) +
  geom_line(size = 1.5) +
  geom_line(data = z.score.figure.4.bulk.RNA.seq.candidates.canonical.ensembl %>% 
              filter(media %in% c("FM", "NM")), aes(stage, value, group = geneName, colour = media), size = 1.5) +
  scale_color_manual(values = c("FM" = "black", "NM" = "#0000FF", "PM" = "#FF8C00"), name = "Culture Media") +
  facet_wrap(~geneName, nrow = 8, ncol = 1) +
  ylab("Gene expression (z-score)") +
  xlab("Reprogramming stages") +
  gplot.theme
ggsave(plot = selected.genes.line.plots,
  "~/projects/Polo_Group/Ethan_Liu/Ethan_Liu_ATAC-seq_Human_Reprogramming_human_samples/analysis/R/plots/Figure_PDFs/Figure_4_bulk_RNA-seq_over_ATAC-seq_changing_clusters_line_plots_reordered.pdf",
  height = 15,
  width = 4.5,
  units = "cm",
  dpi = 300,
  scale = 1
)
```


# Motif enrichment.
Top 25 most enriched motifs (based on p-value) used (`data/bkATAC_motif_enrichment.Rdata`)